File: api/models/data/accounting.py
from django.db import models
from api.models.data.base import BaseModel
from api.models.data.choices import CURRENCY_CHOICES, DEFAULT_CURRENCY


class AccountType(models.TextChoices):
    ASSET = 'asset', 'Asset'
    LIABILITY = 'liability', 'Liability'
    EQUITY = 'equity', 'Equity'
    INCOME = 'income', 'Income'
    EXPENSE = 'expense', 'Expense'


class AccountCategory(BaseModel):
    """Category model for organizing accounts (e.g., Assets, Liabilities, Income, Expenses)"""
    name = models.CharField(max_length=100)
    code = models.CharField(max_length=10, unique=True)
    account_type = models.CharField(max_length=20, choices=AccountType.choices)
    description = models.CharField(max_length=200, blank=True, null=True)

    class Meta:
        ordering = ['code']

    def __str__(self):
        return f"{self.code} - {self.name}"


class Account(BaseModel):
    """Chart of Accounts - Each account in the double-entry system"""
    name = models.CharField(max_length=200)
    code = models.CharField(max_length=20, unique=True)
    category = models.ForeignKey(AccountCategory, on_delete=models.PROTECT, related_name='accounts')
    parent = models.ForeignKey('self', on_delete=models.SET_NULL, null=True, blank=True, related_name='children')
    is_active = models.BooleanField(default=True)
    is_detail = models.BooleanField(default=True, help_text="Can post entries to this account")
    balance = models.DecimalField(max_digits=15, decimal_places=2, default=0)
    currency = models.CharField(max_length=3, choices=CURRENCY_CHOICES, default=DEFAULT_CURRENCY)

    class Meta:
        ordering = ['code']

    def __str__(self):
        return f"{self.code} - {self.name}"

    def get_balance(self):
        """Calculate current balance from journal entries"""
        from django.db.models import Sum
        debits = self.journal_entries.aggregate(total=Sum('debit'))['total'] or 0
        credits = self.journal_entries.aggregate(total=Sum('credit'))['total'] or 0

        # Asset/Expense: Debit increases, Credit decreases
        if self.category.account_type in ['asset', 'expense']:
            return debits - credits
        # Liability/Equity/Income: Credit increases, Debit decreases
        return credits - debits


class JournalEntry(BaseModel):
    """Individual journal entry line (double-entry bookkeeping)"""
    date = models.DateField()
    account = models.ForeignKey(Account, on_delete=models.PROTECT, related_name='journal_entries')
    debit = models.DecimalField(max_digits=15, decimal_places=2, default=0)
    credit = models.DecimalField(max_digits=15, decimal_places=2, default=0)
    description = models.TextField(blank=True, null=True)
    reference = models.CharField(max_length=100, blank=True, null=True)
    transaction = models.ForeignKey('Transaction', on_delete=models.CASCADE, related_name='entries')

    class Meta:
        ordering = ['-date', '-id']

    def __str__(self):
        return f"{self.date} - {self.account.code} - Dr: {self.debit}, Cr: {self.credit}"

    def save(self, *args, **kwargs):
        super().save(*args, **kwargs)
        # Update account balance
        self.account.balance = self.account.get_balance()
        self.account.save(update_fields=['balance'])


class Transaction(BaseModel):
    """A complete transaction with multiple journal entries"""
    TRANSACTION_TYPES = [
        ('student_payment', 'Student Payment'),
        ('expense', 'Expense'),
        ('payroll', 'Payroll'),
        ('advance', 'Advance Salary'),
        ('rental_income', 'Rental Income'),
        ('other_income', 'Other Income'),
        ('journal', 'Manual Journal Entry'),
        ('opening', 'Opening Balance'),
    ]

    number = models.CharField(max_length=50, unique=True, blank=True)
    date = models.DateField()
    description = models.TextField(blank=True, null=True)
    transaction_type = models.CharField(max_length=20, choices=TRANSACTION_TYPES, default='journal')
    reference = models.CharField(max_length=100, blank=True, null=True)
    is_posted = models.BooleanField(default=True)
    posted_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        ordering = ['-date', '-id']

    def __str__(self):
        return f"{self.number} - {self.date} - {self.get_transaction_type_display()}"

    def save(self, *args, **kwargs):
        if not self.number:
            # Generate transaction number
            from django.utils import timezone
            import time
            prefix = self.transaction_type[:3].upper()
            # Use timestamp to ensure uniqueness
            timestamp = int(time.time() * 1000000) % 1000000
            self.number = f"{prefix}-{timezone.now().year}-{timestamp:06d}"
        super().save(*args, **kwargs)

    @property
    def total_debit(self):
        return sum(entry.debit for entry in self.entries.all())

    @property
    def total_credit(self):
        return sum(entry.credit for entry in self.entries.all())

    def is_balanced(self):
        """Check if transaction is balanced (debits = credits)"""
        from decimal import Decimal
        return abs(self.total_debit - self.total_credit) < Decimal('0.01')


class FiscalYear(BaseModel):
    """Fiscal year management"""
    name = models.CharField(max_length=50)
    start_date = models.DateField()
    end_date = models.DateField()
    is_closed = models.BooleanField(default=False)

    class Meta:
        ordering = ['-start_date']

    def __str__(self):
        return self.name

File: frontend/src/types/accounting.ts
export interface AccountCategory {
  id: string;
  name: string;
  code: string;
  account_type: string;
  account_type_display?: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
}

export interface Account {
  id: string;
  name: string;
  code: string;
  category: string;
  category_name?: string;
  account_type?: string;
  parent?: string;
  parent_name?: string;
  is_active: boolean;
  is_detail: boolean;
  balance: number;
  current_balance?: number;
  total_debit?: number;
  total_credit?: number;
  currency?: string;
  created_at?: string;
  updated_at?: string;
}

export interface JournalEntry {
  id: string;
  date: string;
  account: string;
  account_code?: string;
  account_name?: string;
  debit: number;
  credit: number;
  description?: string;
  reference?: string;
  transaction: string;
  transaction_number?: string;
  created_at?: string;
  updated_at?: string;
}

export interface Transaction {
  id: string;
  number: string;
  date: string;
  description?: string;
  transaction_type: string;
  reference?: string;
  is_posted: boolean;
  posted_at?: string;
  total_debit?: number;
  total_credit?: number;
  is_balanced?: boolean;
  entries?: JournalEntry[];
  created_at?: string;
  updated_at?: string;
}

export interface TransactionFormData {
  date: string;
  description?: string;
  transaction_type: string;
  reference?: string;
  entries: JournalEntryFormData[];
}

export interface JournalEntryFormData {
  account: string;
  debit: number;
  credit: number;
  description?: string;
  reference?: string;
}

export interface FiscalYear {
  id: string;
  name: string;
  start_date: string;
  end_date: string;
  is_closed: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface TrialBalance {
  date: string;
  accounts: TrialBalanceAccount[];
  total_debit: number;
  total_credit: number;
  is_balanced: boolean;
}

export interface TrialBalanceAccount {
  code: string;
  name: string;
  type: string;
  debit: number;
  credit: number;
}

export interface IncomeStatement {
  start_date: string;
  end_date: string;
  income: IncomeItem[];
  total_income: number;
  expenses: ExpenseItem[];
  total_expenses: number;
  net_income: number;
  is_profit: boolean;
}

export interface IncomeItem {
  code: string;
  name: string;
  amount: number;
}

export interface ExpenseItem {
  code: string;
  name: string;
  amount: number;
}

export interface BalanceSheet {
  date: string;
  assets: BalanceSheetItem[];
  total_assets: number;
  liabilities: BalanceSheetItem[];
  total_liabilities: number;
  equity: BalanceSheetItem[];
  total_equity: number;
  total_liabilities_and_equity: number;
  is_balanced: boolean;
}

export interface BalanceSheetItem {
  code: string;
  name: string;
  amount: number;
}
